File: examples/go/angzarr/projector_handler.go
package angzarr

import (
	"context"

	angzarrpb "angzarr/proto/angzarr"

	"google.golang.org/grpc"
)

// ProjectorHandleFunc processes an EventBook and returns a Projection.
type ProjectorHandleFunc func(book *angzarrpb.EventBook) (*angzarrpb.Projection, error)

// ProjectorHandler implements the gRPC Projector service.
//
// Projectors consume events and produce read models. The handler receives
// EventBooks and delegates to a user-provided function for projection logic.
type ProjectorHandler struct {
	angzarrpb.UnimplementedProjectorServer
	name     string
	domains  []string
	handleFn ProjectorHandleFunc
}

// NewProjectorHandler creates a projector handler with the given name and input domains.
func NewProjectorHandler(name string, domains ...string) *ProjectorHandler {
	return &ProjectorHandler{name: name, domains: domains}
}

// WithHandle sets the event handling callback.
func (h *ProjectorHandler) WithHandle(fn ProjectorHandleFunc) *ProjectorHandler {
	h.handleFn = fn
	return h
}

// GetDescriptor returns the component descriptor.
func (h *ProjectorHandler) GetDescriptor(_ context.Context, _ *angzarrpb.GetDescriptorRequest) (*angzarrpb.ComponentDescriptor, error) {
	desc := h.Descriptor()
	var inputs []*angzarrpb.Subscription
	for _, input := range desc.Inputs {
		inputs = append(inputs, &angzarrpb.Subscription{
			Domain:     input.Domain,
			EventTypes: input.EventTypes,
		})
	}
	return &angzarrpb.ComponentDescriptor{
		Name:          desc.Name,
		ComponentType: desc.ComponentType,
		Inputs:        inputs,
	}, nil
}

// Handle processes an EventBook and returns a Projection.
func (h *ProjectorHandler) Handle(_ context.Context, book *angzarrpb.EventBook) (*angzarrpb.Projection, error) {
	if h.handleFn != nil {
		return h.handleFn(book)
	}
	return &angzarrpb.Projection{}, nil
}

// Descriptor builds a component descriptor from registered domains.
func (h *ProjectorHandler) Descriptor() Descriptor {
	var inputs []SubscriptionDesc
	for _, domain := range h.domains {
		inputs = append(inputs, SubscriptionDesc{Domain: domain})
	}
	return Descriptor{
		Name:          h.name,
		ComponentType: ComponentProjector,
		Inputs:        inputs,
	}
}

// RunProjectorServer starts a gRPC server for a projector.
func RunProjectorServer(cfg ServerConfig, handler *ProjectorHandler) error {
	return RunServer(cfg, func(s *grpc.Server) {
		angzarrpb.RegisterProjectorServer(s, handler)
	})
}

File: examples/python/angzarr/projector_handler.py
"""ProjectorHandler: gRPC Projector servicer.

Projectors consume events and produce read models. The handler receives
EventBooks and delegates to a user-provided function for projection logic.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Callable

import grpc

from angzarr import projector_pb2_grpc
from angzarr import types_pb2 as types
from router import Descriptor, SubscriptionDesc
from server import run_server

if TYPE_CHECKING:
    import structlog

ProjectorHandleFunc = Callable[[types.EventBook], types.Projection]


class ProjectorHandler(projector_pb2_grpc.ProjectorServicer):
    """gRPC Projector servicer backed by a handle function."""

    def __init__(self, name: str, *domains: str) -> None:
        self._name = name
        self._domains = list(domains)
        self._handle_fn: ProjectorHandleFunc | None = None

    def with_handle(self, fn: ProjectorHandleFunc) -> ProjectorHandler:
        """Set the event handling callback."""
        self._handle_fn = fn
        return self

    def GetDescriptor(
        self,
        request: types.GetDescriptorRequest,
        context: grpc.ServicerContext,
    ) -> types.ComponentDescriptor:
        """Return the component descriptor."""
        desc = self.descriptor()
        return types.ComponentDescriptor(
            name=desc.name,
            component_type=desc.component_type,
            inputs=[
                types.Subscription(domain=inp.domain, event_types=inp.event_types)
                for inp in desc.inputs
            ],
        )

    def Handle(
        self,
        request: types.EventBook,
        context: grpc.ServicerContext,
    ) -> types.Projection:
        """Process an EventBook and return a Projection."""
        if self._handle_fn is not None:
            return self._handle_fn(request)
        return types.Projection()

    def descriptor(self) -> Descriptor:
        """Build a component descriptor."""
        return Descriptor(
            name=self._name,
            component_type="projector",
            inputs=[SubscriptionDesc(domain=d) for d in self._domains],
        )


def run_projector_server(
    name: str,
    default_port: str,
    handler: ProjectorHandler,
    logger: structlog.BoundLogger | None = None,
) -> None:
    """Start a gRPC server for a projector."""
    run_server(
        projector_pb2_grpc.add_ProjectorServicer_to_server,
        handler,
        service_name="Projector",
        domain=name,
        default_port=default_port,
        logger=logger,
    )
